new/resources-anywhere.md

```diff
@@ -0,0 +1,211 @@
+---
+RFC: unassigned
+Author: John Keiser <john@johnkeiser.com>
+Status: Draft
+Type: Standards Track
+---
+
+# Allow Resources In Any Namespace, Remove Method Missing
+
+This proposal addresses:
+- Allowing users to place Chef resources and providers anywhere in their tree, not just `Chef::Resource` and `Chef::Provider`
+- Eliminating the issue where resource and provider code that tries to access `File.x` actually get the File resource, having to type `::File` instead to get what they want.
+- Eliminating `method_missing` as a way to look up resources, in favor of explicit DSL (which gets rid of a great deal of confusing error message).
+
+## Motivation
+
+    As a <<user_profile>>,
+    I want to <<functionality>>,
+    so that <<benefit>>.
+
+## Specification
+
+### Resources outside of `Chef::Resource`
+
+Users who wish to register a resource named `my_resource` can presently do it by creating a class inside the `Chef::Resource` namespace and another in `Chef::Provider`.  Chef then uses magical namespace lookups to
+
+```ruby
+# chef/resource/
+class Chef::Resource::MyResource < Chef::Resource
+end
+```
+
+With this proposal, the above will still work; but users can now create resources outside of Chef::Resource with the following classes:
+
+```ruby
+class MyResource < Chef::Resource
+  ...
+  class Provider < Chef::Provider
+    ...
+  end
+end
+```
+
+The mechanism is:
+- An `inherited` hook on `Chef::Resource` automatically takes care of hooking up `MyResource` to all recipe DSL when the class is created.
+- The provider lookup for a resource looks first for a `Provider` class inside the Resource before moving on to other things.
+
+### Accessing `File`
+
+Accessing top-level objects like `File` are one of the big reasons for this proposal.  Anything that extends from `Chef::Resource` *will* look up `Chef::Resource::X` before `::X` when a method types `X`.
+
+To fix this, we will move resources and providers in `Chef::Provider::*` and `Chef::Resource::*` into `Chef::Resources` (and use the inline syntax as well).  For addressing the obvious concerns, see the Errors and Deprecations section.
```

Comment:
This will break a ton of people.  The chef classes are commonly inherited from and the provider parameter to resources exposes the class name as a public API (highly unfortunately).  We probably should rewrite the existing chef classes as new resources, but we probably need to keep the existing class layout.  The disruption from changing that would likely be of the order of the Chef 11 attributes changes or maybe worse.  Making those changes just to fix `::File` (which is fixed in a lot of use cases like in not_if blocks) probably isn't compelling enough to break everything this badly.